diff --git a/environment.h b/environment.h
index 27f657a..3c8940f 100644
--- a/environment.h
+++ b/environment.h
@@ -159,6 +159,10 @@ extern int zlib_compression_level;
 extern int pack_compression_level;
 extern unsigned long pack_size_limit_cfg;
 
+#ifdef __MVS__
+extern int utf8_ccsid;
+extern int iconv_translit;
+#endif
 extern int precomposed_unicode;
 extern int protect_hfs;
 extern int protect_ntfs;
