diff --git a/builtin/index-pack.c b/builtin/index-pack.c
index b67fb02..1918ff3 100644
--- a/builtin/index-pack.c
+++ b/builtin/index-pack.c
@@ -212,6 +212,9 @@ static void init_thread(void)
 	CALLOC_ARRAY(thread_data, nr_threads);
 	for (i = 0; i < nr_threads; i++) {
 		thread_data[i].pack_fd = xopen(curr_pack, O_RDONLY);
+#ifdef __MVS__
+		__disableautocvt(thread_data[i].pack_fd);
+#endif
 	}
 
 	threads_active = 1;
@@ -368,10 +371,17 @@ static const char *open_pack_file(const char *pack_name)
 			pack_name = strbuf_detach(&tmp_file, NULL);
 		} else {
 			output_fd = xopen(pack_name, O_CREAT|O_EXCL|O_RDWR, 0600);
+#ifdef __MVS__
+			__setfdbinary(output_fd);
+			__disableautocvt(output_fd);
+#endif
 		}
 		nothread_data.pack_fd = output_fd;
 	} else {
 		input_fd = xopen(pack_name, O_RDONLY);
+#ifdef __MVS__
+		__disableautocvt(input_fd);
+#endif
 		output_fd = -1;
 		nothread_data.pack_fd = input_fd;
 	}
